src/expr/proof_node_updater.cpp

```diff
@@ -0,0 +1,82 @@
+/*********************                                                        */
+/*! \file proof_node_updater.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Morgan Deters, Andrew Reynolds, Tim King
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of a utility for updating proof nodes
+ **/
+
+#include "expr/proof_node_updater.h"
+
+#include "expr/lazy_proof.h"
+
+using namespace CVC4::kind;
+
+namespace CVC4 {
+
+ProofNodeUpdaterCallback::ProofNodeUpdaterCallback() {}
+ProofNodeUpdaterCallback::~ProofNodeUpdaterCallback() {}
+
+ProofNodeUpdater::ProofNodeUpdater(ProofNodeManager* pnm,
+                                   ProofNodeUpdaterCallback& cb)
+    : d_pnm(pnm), d_cb(cb)
+{
+}
+
+void ProofNodeUpdater::process(std::shared_ptr<ProofNode> pf)
+{
+  Trace("pf-process") << "ProofNodeUpdater::process" << std::endl;
+  std::unordered_set<ProofNode*> visited;
+  std::unordered_set<ProofNode*>::iterator it;
+  std::vector<ProofNode*> visit;
+  ProofNode* cur;
+  visit.push_back(pf.get());
+  do
+  {
+    cur = visit.back();
+    it = visited.find(cur);
+    if (it == visited.end())
+    {
+      visited.insert(cur);
+      // should it be updated?
+      if (d_cb.shouldUpdate(cur))
+      {
+        PfRule id = cur->getRule();
+        LazyCDProof lcp(d_pnm);
```

Comment:
say why lazy proof

Reply:
Didn't have to be lazy proof, I've updated.